crypto/cipher/aes_gcm_nss.c

```diff
@@ -339,21 +343,24 @@ static srtp_err_status_t srtp_aes_gcm_nss_encrypt(void *cv,
     // where it can write the tag.  We can't just use c->tag because
     // memcpy has undefined behavior on overlapping ranges.
     uint8_t tagbuf[16];
-    uint8_t *non_null_buf = buf;
-    if (!non_null_buf && (*enc_len == 0)) {
+    const uint8_t *non_null_buf = src;
+    uint8_t *non_null_dst_buf = dst;
+    if (!non_null_buf && (src_len == 0)) {
         non_null_buf = tagbuf;
+        non_null_dst_buf = tagbuf;
+        *dst_len = sizeof(tagbuf);
     } else if (!non_null_buf) {
         return srtp_err_status_bad_param;
     }
 
-    srtp_err_status_t status =
-        srtp_aes_gcm_nss_do_crypto(cv, true, non_null_buf, enc_len);
+    srtp_err_status_t status = srtp_aes_gcm_nss_do_crypto(
+        cv, true, non_null_buf, src_len, non_null_dst_buf, dst_len);
     if (status != srtp_err_status_ok) {
         return status;
     }
 
-    memcpy(c->tag, non_null_buf + (*enc_len - c->tag_size), c->tag_size);
-    *enc_len -= c->tag_size;
+    memcpy(c->tag, non_null_dst_buf + (*dst_len - c->tag_size), c->tag_size);
+    *dst_len -= c->tag_size;
```

Comment:
This sets the length in the formal param `dst_len` to be whatever it was coming in minus the tag size.  What if `dst_size` was 0 or 1, for example?  Since I cannot see any data coming out when the `src` buffer is NULL, is this just indicating how much buffer would remain if a tag had been produced?  (I understand a tag is somehow produced for something and stored in `c->tag`, but I don't see the relationship with that `dst` parameter when the `src` is NULL.)

Reply:
you are right that this is confusing and if we do #714 "consider merging the gcm encrypt and get tag functions of cipher api" once this is merged it will clean all this up alot.

As it is now though, if the call to srtp_aes_gcm_nss_do_crypto() succeeds then dst_len was large enough for the tag and any encrypted data, it then would have been update to be that size. Since currently the tag is not going to be returned from this function it is truncated by tag size. But I agree that a sanity check against tag size would have be good here.